libcontainer/cgroups/devices/systemd.go

```diff
@@ -239,3 +239,43 @@ func allowAllDevices() []systemdDbus.Property {
 		newProp("DeviceAllow", []deviceAllowEntry{}),
 	}
 }
+
+func getCharEntryPath(rule *devices.Rule) string {
+	if isNVIDIADevice(rule) {
+		return getNVIDIAEntryPath(rule)
+	}
+	return fmt.Sprintf("/dev/char/%d:%d", rule.Major, rule.Minor)
+}
+
+func isNVIDIADevice(rule *devices.Rule) bool {
+	// NVIDIA device has major 195 and 507
+	if rule.Major == 195 || rule.Major == 507 {
+		return true
+	}
+	return false
+}
+
+func getNVIDIAEntryPath(rule *devices.Rule) string {
+	path := "/dev"
+	switch rule.Major {
+	case 195:
+		switch rule.Minor {
+		case 254:
+			path = fmt.Sprintf("%s/%s", path, "nvidia-modeset")
+		case 255:
+			path = fmt.Sprintf("%s/%s", path, "nvidiactl")
+		default:
+			// nvidia0 /dev/nvidia0
+			// nvidia1 /dev/nvidia1
+			path = fmt.Sprintf("%s/%s%d", path, "nvidia", rule.Minor)
+		}
+	case 507:
+		switch rule.Minor {
+		case 0:
+			path = fmt.Sprintf("%s/%s", path, "nvidia-uvm")
+		case 1:
+			path = fmt.Sprintf("%s/%s", path, "nvidia-uvm-tools")
+		}
+	}
+	return path
```

Comment:
We should never hit this case, because the code currently checks if `isNVIDIADevice` is true,  but if this function would ever be called from somewhere else, and Major is neither `195` or `507`, which would return `/dev`.

Perhaps it's better to combine the functions, to make sure we never can hit that situation, i.e. this could be in `getCharEntryPath()` itself;


```go
func getCharEntryPath(rule *devices.Rule) string {
	switch rule.Major {
	case 195:
		// Special case for NVIDIA devices
		switch rule.Minor {
		case 254:
			return "/dev/nvidia-modeset"
		case 255:
			return "/dev/nvidiactl"
		default:
			// nvidia0 /dev/nvidia0
			// nvidia1 /dev/nvidia1
			return fmt.Sprintf("dev/nvidia%d", rule.Minor)
		}
	case 507:
		// Special case for NVIDIA devices
		switch rule.Minor {
		case 0:
			return "/dev/nvidia-uvm"
		case 1:
			return "/dev/nvidia-uvm-tools"
		}
	}

	return fmt.Sprintf("/dev/char/%d:%d", rule.Major, rule.Minor)
}
```